plugins/hotkeyshortcuts.koplugin/main.lua

```diff
@@ -0,0 +1,574 @@
+    local cursor_keys = { "Up", "Down", "Left", "Right" }
+    local page_turn_keys = { "LPgBack", "LPgFwd", "RPgBack", "RPgFwd" }
+    local function_keys = { "Back", "Home", "Press", "Menu" }
+    local key_name_mapping = {
+        LPgBack = "left_page_back",    RPgBack = "right_page_back",
+        LPgFwd  = "left_page_forward", RPgFwd  = "right_page_forward",
+    }
+    local function addKeyEvent(modifier, key, event, args)
+        self.key_events[modifier .."Plus".. key] = { { modifier, key }, event = event, args = args }
+    end
+
+    local function addKeyEvents(modifier, keys, event, args_prefix)
+        for _, key in ipairs(keys) do
+            local mapped_key = key_name_mapping[key] or key:lower()
+            addKeyEvent(modifier, key, event, args_prefix .. mapped_key)
+        end
+    end
+
+    local modifier = Device:hasScreenKB() and "ScreenKB" or "Shift"
+    addKeyEvents(modifier, cursor_keys, "HotkeyAction", "modifier_plus_")
+    if not self.is_docless then
+        addKeyEvents(modifier, page_turn_keys, "HotkeyAction", "modifier_plus_")
+        addKeyEvent(modifier, "Press", "HotkeyAction", "modifier_plus_press")
+        if self.settings_data.data["press_key_does_hotkeyshortcuts"] then
+            self.key_events.Press = { { "Press" }, event = "HotkeyAction", args = "press" }
+        end
+    end
+    addKeyEvent(modifier, "Back", "HotkeyAction", "modifier_plus_back")
+    addKeyEvent(modifier, "Home", "HotkeyAction", "modifier_plus_home")
+    -- remember, screenkb+menu is already used for screenshots (on k4), don't add it here.
+
+    if Device:hasKeyboard() then
+        addKeyEvent("Shift", "Menu", "HotkeyAction", "modifier_plus_menu")
+        local second_modifier = Device:hasSymKey() and "Alt" or "Ctrl"
+        addKeyEvents(second_modifier, cursor_keys, "HotkeyAction", "alt_plus_")
+        addKeyEvents(second_modifier, page_turn_keys, "HotkeyAction", "alt_plus_")
+        addKeyEvents(second_modifier, function_keys, "HotkeyAction", "alt_plus_")
+        local top_row_keys = { "Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P" }
+        local remaining_keys = { "A", "S", "D", "F", "G", "H", "J", "K", "L", "Z", "X", "C", "V", "B", "N", "M" }
+        if Device.k3_alt_plus_key_kernel_translated then
+            -- Add the infamous top row keys, with kernel issues
+            for _, key in ipairs(top_row_keys) do
+                self.key_events["AltPlus" .. key] = {
+                    { Device.k3_alt_plus_key_kernel_translated[key] },
+                    event = "HotkeyAction",
+                    args = "alt_plus_" .. key:lower()
+                }
+            end
+        else
+            addKeyEvents(second_modifier, top_row_keys, "HotkeyAction", "alt_plus_")
+        end
+        addKeyEvents(second_modifier, remaining_keys, "HotkeyAction", "alt_plus_")
+    end -- if hasKeyboard()
+
+    if next(self.key_events) == nil then
+        logger.warn("No hotkey events registered.")
```

Comment:
I meant:
If it can happen because the user sets no shortcut, it should be logger.dbg().
If it can't happen because there's always a shortcut set, you don't need it.
(Anything can happen while you dev, and you can add all the logger.warn() you need - but once you're done, nothing unexpected should happen :))

Reply:
but tommorow Boris Dickie might come and mangle something somewhere that breaks how keys work and suddenly little Timmy's key events won't work and he'll look at the logs and see what is going on...